src/Accounts/Authentication/KeyStore/StorageHelper.cs

```diff
@@ -0,0 +1,282 @@
+        {
+            if (keystore == null)
+            {
+                throw new ArgumentNullException(nameof(keystore));
+            }
+
+            _logger.LogInformation($"Registering token cache with on disk storage");
+
+            keystore.SetOnUpdate(WriteToCachedStorage);
+
+            _logger.LogInformation($"Done initializing");
+        }
+
+        public void UnregisterCache(IKeyStore keystore)
+        {
+            if (keystore == null)
+            {
+                throw new ArgumentNullException(nameof(keystore));
+            }
+            keystore.SetOnUpdate(null);
+        }
+
+        public void LoadFromCachedStorage(IKeyStore keystore)
+        {
+            LogMessage(EventLogLevel.Verbose, $"Before access\nAcquiring lock for token cache");
+
+            using (CreateCrossPlatLock(_storageCreationProperties))
+            {
+                LogMessage(EventLogLevel.Verbose, $"Before access, the store has changed");
+
+                byte[] cachedStoreData = null;
+                try
+                {
+                    cachedStoreData = PersistanceStore.ReadData();
+                }
+                catch (Exception ex)
+                {
+                    LogMessage(EventLogLevel.Error, $"Could not read the token cache. Ignoring. Exception: {ex}");
+                    return;
+
+                }
+                LogMessage(EventLogLevel.Verbose, $"Read '{cachedStoreData?.Length}' bytes from storage");
+
+                try
+                {
+                    LogMessage(EventLogLevel.Verbose, $"Deserializing the store");
+                    keystore.Deserialize(cachedStoreData, false);
+                }
+                catch (Exception e)
+                {
+                    LogMessage(EventLogLevel.Error, $"An exception was encountered while deserializing the {nameof(StorageHelper)} : {e}");
+                    LogMessage(EventLogLevel.Error, $"No data found in the store, clearing the cache in memory.");
+
+                    PersistanceStore.Clear();
+                    throw;
+                }
+            }
+        }
+
+        public void WriteToCachedStorage(KeyStoreNotificationArgs args)
+        {
+            using (CreateCrossPlatLock(_storageCreationProperties))
+            {
+                LogMessage(EventLogLevel.Verbose, $"After access");
+                byte[] data = null;
+                LogMessage(EventLogLevel.Verbose, $"After access, cache in memory HasChanged");
+                try
+                {
+                    data = args.KeyStore.Serialize();
+                }
+                catch (Exception e)
+                {
+                    LogMessage(EventLogLevel.Error, $"An exception was encountered while serializing the {nameof(StorageHelper)} : {e}");
+                    LogMessage(EventLogLevel.Error, $"No data found in the store, clearing the cache in memory.");
+
+                    PersistanceStore.Clear();
```

Comment:
Clear calls WriteToCachedStorage recursively.

Reply:
Sorry for the mistake. PersistanceStore clear won't call WriteToCachedStorage recsusively. It's InMemoryKeyStore clear will do.